src/model/styles.rs

```diff
@@ -281,9 +294,100 @@ impl Selector {
         Self::Can(TypeId::of::<T>())
     }
 
+    /// Matches the selector for an introspector.
+    pub fn match_iter<'a>(
+        &'a self,
+        introspector: &'a Introspector,
+    ) -> Box<dyn Iterator<Item = (usize, &'a Content)> + 'a> {
+        self.match_iter_inner(introspector, introspector.all().enumerate())
+    }
+
+    /// Transforms this selector into a [`Selector::Before`] selector.
+    pub fn before(self, location: Location, inclusive: bool) -> Self {
+        Self::Before(Box::new(self), location, inclusive)
+    }
+
+    /// Transforms this selector into a [`Selector::After`] selector.
+    pub fn after(self, location: Location, inclusive: bool) -> Self {
+        Self::After(Box::new(self), location, inclusive)
+    }
+
+    /// Transforms this selector and an iterator of other selectors into a
+    /// [`Selector::Or`] selector.
+    pub fn and(self, others: impl IntoIterator<Item = Self>) -> Self {
+        Self::And(others.into_iter().chain(Some(self)).collect())
+    }
+
+    /// Transforms this selector and an iterator of other selectors into a
+    /// [`Selector::And`] selector.
+    pub fn or(self, others: impl IntoIterator<Item = Self>) -> Self {
+        Self::Or(others.into_iter().chain(Some(self)).collect())
+    }
+
+    /// Match the selector against the given list of elements.
+    /// Returns an iterator over the matching elements.
+    fn match_iter_inner<'a>(
```

Comment:
is the purpose of this only to speed up lookup of `location` selectors? or are there other advantages over a simple loop with `matches`?

Reply:
The purpose is to make the before and after work better by using iterators (and therefore only looking up location once) rather than by checking indices every iterations.